ste/concurrencyTuner.go

```diff
@@ -0,0 +1,275 @@
+// Copyright © Microsoft <wastore@microsoft.com>
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package ste
+
+import (
+	"github.com/Azure/azure-storage-azcopy/common"
+	"sync"
+	"time"
+)
+
+type ConcurrencyTuner interface {
+	// GetRecommendedConcurrency is called repeatedly, at intervals decided by the caller,
+	// to compute recommended concurrency levels
+	GetRecommendedConcurrency(currentMbps int, highCpuUsage bool) (newConcurrency int, reason string)
+
+	// RequestCallbackWhenStable lets interested parties ask the concurrency tuner to call them back when the tuner has reached a stable level
+	RequestCallbackWhenStable(callback func()) (callbackAccepted bool)
+
+	// GetFinalState returns the final state of the tuner
+	GetFinalState() (finalReason string, finalRecommendedConcurrency int, tm time.Time)
+}
+
+type nullConcurrencyTuner struct {
+	fixedValue int
+}
+
+func (n *nullConcurrencyTuner) GetRecommendedConcurrency(currentMbps int, highCpuUsage bool) (newConcurrency int, reason string) {
+	return n.fixedValue, concurrencyReasonFinished
+}
+
+func (n *nullConcurrencyTuner) RequestCallbackWhenStable(callback func()) (callbackAccepted bool) {
+	return false
+}
+
+func (n *nullConcurrencyTuner) GetFinalState() (finalReason string, finalRecommendedConcurrency int, tm time.Time) {
+	return concurrencyReasonTunerDisabled, 0, time.Time{}
+}
+
+type autoConcurrencyTuner struct {
```

Comment:
Not super sure where to put this, but perhaps we could cache the final value of this somewhere so a user wouldn't have to re-tune every time?

Reply:
Yeah, that's tempting... Tricky part is not just "where do we cache it", but also the fact that the correct value depends on:
* file size
* target account type
* target region
* other network traffic (and therefore, potentially, time-of-day)
* S2S vs upload/download

So there's no one-size fits all answer.  I think, long term, it could be nice to make the auto-tuning the default, but we should see how auto-tuning performs in benchmark mode first before making it the default.  If we make it the default, we may also want to do a variation of your suggestion - rather than cache an exact value, cache the starting value for the auto-tuning. So that time is not wasted on low values, like 8 or 16, in cases where we know that it will end up tuning to something much higher.